File: djsteevbot.js
/* Copyright 2013 Steven Pledger <steev@yourewinner.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
var Bot    = require('ttapi'), repl = require('repl');
var AUTH   = 'XXXXXXXXXXXXXXXXXXXXXXXX';
var USERID = 'XXXXXXXXXXXXXXXXXXXXXXXX';
var ROOMID = 'XXXXXXXXXXXXXXXXXXXXXXXX';

var bot = new Bot(AUTH, USERID, ROOMID);
var djs = {};

repl.start('> ').context.bot = bot;

bot.on('speak', function (data) {
	// Get the data
	var name = data.name;
	var text = data.text;
	
	if (text.match(/^\/join$/)) {
		bot.addDj();
	} else if (text.match(/^\/leave$/)) {
		bot.remDj();
	} else if (text.match(/^\/skip$/)) {
		bot.skip();
	} else if (text.match(/^\/snag$/)) {
		bot.roomInfo(function(data) {
			var songid = data.room.metadata.current_song._id;
			var song = data.room.metadata.current_song.metadata.song;
			bot.snag();
			bot.playlistAll(function(playlist) {
				// Use the length so that it gets added to the end of the list.
				bot.playlistAdd(songid, playlist.list.length); 
			});
			bot.speak('Added '+song+' to the playlist.');
		});
	} else if (text.match(/^\/yw$/)) {
		bot.speak("You're Winner!");
	} else if (text.match(/should +i +play/i)) {
		bot.speak('@'+name+' play BACK THAT AZZ UP!');
	} else if (text.match(/bop/)) {
		bot.speak('FUCK YEAH I LOVE THIS SONG!!! <3');
		bot.bop();
	} else if (text.match(/turd/)) {
		bot.speak("BOO THIS SHIT SUCKS!!!");
		bot.vote('down');
	} else if (text.match(/^\/kickme [0-9]+$/)) {
		var djId = data.userid;
		var num = parseInt(text.split(" ")[1]);
		if (djs[djId]) {
			djs[djId].nbSong = 0; // reset plays
			djs[djId].songsLimit = num;
			bot.speak('@'+name+' I will remove you after '+num+' plays.');
		}
	} else if (text.match(/^\/setlaptop .+$/)) {
		var type = text.split(" ")[1];
		bot.modifyLaptop(type);
	} else if (text.match(/^\/setavatar [0-9]+$/)) {
		num = parseInt(text.split(" ")[1]);
		bot.setAvatar(num);
	} else if (text.match(/^\/bootuser .+$/)) {
		if (name == "djsteev") {
			var user = text.split(" ")[1];
			bot.roomInfo(function (data) {
				for (var i = 0; i < data.users.length; i++) {
					if (data.users[i].name == user) {
						bot.bootUser(data.users[i].userid);
						break;
					}
				}
			});
		} else {
			bot.speak("Only steev can do that. Asshole");
		}
	}

});

bot.on('endsong', function (data) {
	var room = data.room;
	var song = room.metadata.current_song.metadata.song;
	var upvotes = room.metadata.upvotes;
	var downvotes = room.metadata.downvotes;
	bot.speak(':musical_note: '+song+' :thumbsup: '+upvotes+' :thumbsdown: '+downvotes+'');
	
	var djId = data.room.metadata.current_dj;
	if (djs[djId] && djs[djId].songsLimit > 0 && ++djs[djId].nbSong >= djs[djId].songsLimit) {
		bot.remDj(djId);
		delete djs[djId];
	}
});

bot.on('roomChanged', function (data) {
	var currentDjs = data.room.metadata.djs;
	for (var i = 0; i < currentDjs.length; i++) {
		djs[currentDjs[i]] = { nbSong: 0, songsLimit: 0 };
	}
});

bot.on('add_dj', function (data) {
	djs[data.user[0].userid] = { nbSong: 0, songsLimit: 0 };
});

bot.on('rem_dj', function (data) {
	delete djs[data.user[0].userid];
});

File: djsteevbot.py
#!/usr/bin/env python
# Copyright 2013 Steven Pledger <steev@yourewinner.com>
"""
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
from ttapi import Bot
from re import match

AUTH = 'XXXXXXXXXXXXXXXXXXXXXXXX'
USERID = 'XXXXXXXXXXXXXXXXXXXXXXXX'
ROOMID = 'XXXXXXXXXXXXXXXXXXXXXXXX'

bot = Bot(AUTH, USERID, ROOMID)
djs = {}

# Callbacks
def onSpeak(data):
	name = data['name']
	text = data['text']
	
	if text == '/join':
		bot.addDj()
	elif text == '/leave':
		bot.remDj()
	elif text == '/rejoin':
		bot.remDj()
		bot.addDj()
	elif text == '/skip':
		bot.skip()
	elif text == '/snag':
		bot.roomInfo(snag)
	elif text == '/yw':
		bot.speak("@" + name + " YOU'RE WINNER !")
	elif 'should i play' in text.lower():
		bot.speak('@' + name + ' play BACK THAT AZZ UP!')
	elif text == '/bop':
		bot.speak('FUCK YEAH I LOVE THIS SONG!!! <3')
		bot.bop()
	elif text == '/turd':
		bot.speak('BOO THIS SHIT SUCKS!!! :thumbsdown:')
		bot.vote('down')
	elif match('^/setlaptop [a-z]+$', text):
		laptop = text.split()[1]
		bot.modifyLaptop(laptop)
	elif match('^/setavatar [0-9]+$', text):
		avatar = int(text.split()[1])
		bot.setAvatar(avatar)
	elif match('^/kick .+$', text):
		if name == 'djsteev':
			name = text.split()[1]
			bot.roomInfo(lambda data: kick(data, name))
		else:
			bot.speak('Only steev can do that. Asshole')
	elif match('^/kickme [0-9]+$', text):
		dj = data['userid']
		num = int(text.split()[1])
		if dj in djs:
			djs[dj]['song_number'] = 0 # reset plays
			djs[dj]['song_limit'] = num
			bot.speak('@' + name + ' I will remove you after ' + str(num) + ' plays.')
	elif text in ('/help', '/commands'):
		bot.speak('DJSteevBot commands: /join, /leave, /rejoin, /skip, /yw, /snag, /bop, /turd, /setlaptop type, /setavatar id, /kickme num, /kick user')

def onEndSong(data):
	song = data['room']['metadata']['current_song']['metadata']['song']
	upvotes = data['room']['metadata']['upvotes']
	downvotes = data['room']['metadata']['downvotes']
	bot.speak(':musical_note: ' + song + ' :thumbsup: ' + str(upvotes) + ' :thumbsdown: ' + str(downvotes))
	
	dj = data['room']['metadata']['current_dj']
	djs[dj]['song_number'] += 1
	if djs[dj]['song_limit'] > 0 and djs[dj]['song_number'] >= djs[dj]['song_limit']:
		bot.remDj(dj)
		del djs[dj]

def onRoomChanged(data):
	currentDjs = data['room']['metadata']['djs']
	for dj in currentDjs:
		djs[dj] = { 'song_number': 0, 'song_limit': 0 }

def onAddDj(data):
	dj = data['user'][0]['userid']
	djs[dj] = { 'song_number': 0, 'song_limit': 0 }

def onRemDj(data):
	dj = data['user'][0]['userid']
	del djs[dj]

def snag(data):
	songid = data['room']['metadata']['current_song']['_id']		
	song = data['room']['metadata']['current_song']['metadata']['song']
	# Add song to the end of the playlist
	bot.playlistAll(lambda playlist: bot.playlistAdd(songid, len(playlist['list'])))
	# show the heart animation
	bot.snag()	
	bot.speak('Added ' + song + ' to the playlist')

def kick(data, name):
	for user in data['users']:
		if user['name'] == name:
			bot.bootUser(user['userid'])
			break

bot.on('speak', onSpeak)
bot.on('endsong', onEndSong)
bot.on('roomChanged', onRoomChanged)
bot.on('add_dj', onAddDj)
bot.on('rem_dj', onRemDj)

bot.start()
